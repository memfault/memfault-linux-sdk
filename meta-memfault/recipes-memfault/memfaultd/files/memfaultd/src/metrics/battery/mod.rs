//
// Copyright (c) Memfault, Inc.
// See License.txt for details
mod battery_reading_handler;
pub use battery_reading_handler::BatteryReadingHandler;

mod battery_monitor;
pub use battery_monitor::BatteryMonitor;
pub use battery_monitor::BatteryMonitorReading;
