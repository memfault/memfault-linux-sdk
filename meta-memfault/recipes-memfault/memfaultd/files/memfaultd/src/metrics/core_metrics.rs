//
// Copyright (c) Memfault, Inc.
// See License.txt for details
// Connectivity metrics
pub const METRIC_MF_SYNC_SUCCESS: &str = "sync_memfault_successful";
pub const METRIC_MF_SYNC_FAILURE: &str = "sync_memfault_failure";
pub const METRIC_CONNECTED_TIME: &str = "connectivity_connected_time_ms";
pub const METRIC_EXPECTED_CONNECTED_TIME: &str = "connectivity_expected_time_ms";
pub const METRIC_SYNC_SUCCESS: &str = "sync_successful";
pub const METRIC_SYNC_FAILURE: &str = "sync_failure";

// Stability metrics
pub const METRIC_OPERATIONAL_HOURS: &str = "operational_hours";
pub const METRIC_OPERATIONAL_CRASHFREE_HOURS: &str = "operational_crashfree_hours";
pub const METRIC_OPERATIONAL_CRASHES: &str = "operational_crashes";

// Battery metrics
pub const METRIC_BATTERY_DISCHARGE_DURATION_MS: &str = "battery_discharge_duration_ms";
pub const METRIC_BATTERY_SOC_PCT_DROP: &str = "battery_soc_pct_drop";
pub const METRIC_BATTERY_SOC_PCT: &str = "battery_soc_pct";
