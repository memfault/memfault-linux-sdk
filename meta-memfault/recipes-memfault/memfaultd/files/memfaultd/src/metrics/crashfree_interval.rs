//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::{
    cmp::max,
    iter::once,
    sync::mpsc::{channel, Receiver, Sender},
    time::Duration,
};

use chrono::Utc;
use tiny_http::{Method, Request, Response};

use crate::{
    http_server::{HttpHandler, HttpHandlerResult},
    metrics::{
        core_metrics::{
            METRIC_OPERATIONAL_CRASHES, METRIC_OPERATIONAL_CRASHFREE_HOURS,
            METRIC_OPERATIONAL_HOURS,
        },
        MetricReading,
    },
    util::time_measure::TimeMeasure,
};

use super::{KeyedMetricReading, MetricStringKey};

pub struct CrashFreeIntervalTracker<T: TimeMeasure> {
    last_interval_mark: T,
    last_crashfree_interval_mark: T,
    crash_count: u32,
    sender: Sender<T>,
    receiver: Receiver<T>,
    interval: Duration,
    elapsed_intervals_key: MetricStringKey,
    crashfree_intervals_key: MetricStringKey,
    crash_count_key: MetricStringKey,
}

#[derive(Debug, PartialEq, Eq)]
struct TimeMod<T: TimeMeasure> {
    count: u32,
    mark: T,
}

impl<T> CrashFreeIntervalTracker<T>
where
    T: TimeMeasure + Copy + Ord + std::ops::Add<Duration, Output = T> + Send + Sync + 'static,
{
    pub fn new(
        interval: Duration,
        elapsed_intervals_key: MetricStringKey,
        crashfree_intervals_key: MetricStringKey,
        crash_count_key: MetricStringKey,
    ) -> Self {
        let (sender, receiver) = channel();
        Self {
            last_crashfree_interval_mark: T::now(),
            last_interval_mark: T::now(),
            sender,
            receiver,
            crash_count: 0,
            interval,
            elapsed_intervals_key,
            crashfree_intervals_key,
            crash_count_key,
        }
    }

    /// Returns a tracker with an hourly interval
    pub fn new_hourly() -> Self {
        Self::new(
            Duration::from_secs(3600),
            METRIC_OPERATIONAL_HOURS.parse().unwrap(),
            METRIC_OPERATIONAL_CRASHFREE_HOURS.parse().unwrap(),
            METRIC_OPERATIONAL_CRASHES.parse().unwrap(),
        )
    }

    /// Wait for the next crash or update the metrics if the wait duration has passed.
    ///
    /// This allows us to have instant updates on crashes and hourly updates on the metrics, but
    /// also allows us to periodically update the metrics so that we don't have to wait for a crash.
    pub fn wait_and_update(&mut self, wait_duration: Duration) -> Vec<KeyedMetricReading> {
        if let Ok(crash_ts) = self.receiver.recv_timeout(wait_duration) {
            // Drain the receiver to get all crashes that happened since the last update
            self.receiver
                .try_iter()
                .chain(once(crash_ts))
                .for_each(|ts| {
                    self.crash_count += 1;
                    self.last_crashfree_interval_mark = max(self.last_crashfree_interval_mark, ts);
                });
        }

        // Since timing out just means no crashes occurred in the `wait_duration`,
        // update even when the receiver times out.
        self.update()
    }

    fn update(&mut self) -> Vec<KeyedMetricReading> {
        let TimeMod {
            count: count_op_interval,
            mark: last_counted_op_interval,
        } = Self::full_interval_elapsed_since(self.interval, &self.last_interval_mark);
        let TimeMod {
            count: count_crashfree_interval,
            mark: last_counted_crashfree_interval,
        } = Self::full_interval_elapsed_since(self.interval, &self.last_crashfree_interval_mark);

        self.last_interval_mark = last_counted_op_interval;
        self.last_crashfree_interval_mark = last_counted_crashfree_interval;

        let crashes = self.crash_count;
        self.crash_count = 0;

        let metrics_ts = Utc::now();
        vec![
            KeyedMetricReading::new(
                self.elapsed_intervals_key.clone(),
                MetricReading::Counter {
                    value: count_op_interval as f64,
                    timestamp: metrics_ts,
                },
            ),
            KeyedMetricReading::new(
                self.crashfree_intervals_key.clone(),
                MetricReading::Counter {
                    value: count_crashfree_interval as f64,
                    timestamp: metrics_ts,
                },
            ),
            KeyedMetricReading::new(
                self.crash_count_key.clone(),
                MetricReading::Counter {
                    value: crashes as f64,
                    timestamp: metrics_ts,
                },
            ),
        ]
    }

    pub fn http_handler(&mut self) -> Box<dyn HttpHandler> {
        Box::new(CrashFreeIntervalHttpHandler {
            channel: self.sender.clone(),
        })
    }

    pub fn capture_crash(&self) {
        self.sender
            .send(T::now())
            .expect("Failed to send crash timestamp");
    }

    /// Count how many `interval` have elapsed since `since`.
    ///
    /// This returns the number of intervals that have elapsed since `since`, and the timestamp of the end of the last interval
    /// that was counted. This is the value you should pass as `since` next time you call this function.
    ///
    /// See unit test for examples.
    fn full_interval_elapsed_since(interval: Duration, since: &T) -> TimeMod<T> {
        let now = T::now();
        if *since > now {
            return TimeMod {
                count: 0,
                mark: T::now(),
            };
        }

        let duration = now.since(since);
        let count_interval_elapsed = (duration.as_nanos() / interval.as_nanos()) as u32;
        TimeMod {
            count: count_interval_elapsed,
            mark: since.add(interval * count_interval_elapsed),
        }
    }
}

struct CrashFreeIntervalHttpHandler<T> {
    channel: Sender<T>,
}

impl<T> HttpHandler for CrashFreeIntervalHttpHandler<T>
where
    T: TimeMeasure + Copy + Ord + std::ops::Add<Duration, Output = T> + Send + Sync,
{
    fn handle_request(&self, request: &mut Request) -> HttpHandlerResult {
        if request.url() == "/v1/crash/report" && request.method() == &Method::Post {
            self.channel.send(T::now()).unwrap();
            HttpHandlerResult::Response(Response::from_string("OK").boxed())
        } else {
            HttpHandlerResult::NotHandled
        }
    }
}

#[cfg(test)]
mod tests {
    use std::time::Duration;

    use rstest::rstest;

    use crate::{
        metrics::{
            crashfree_interval::{
                METRIC_OPERATIONAL_CRASHES, METRIC_OPERATIONAL_CRASHFREE_HOURS,
                METRIC_OPERATIONAL_HOURS,
            },
            KeyedMetricReading, MetricReading,
        },
        test_utils::TestInstant,
    };

    use super::CrashFreeIntervalTracker;
    use super::TimeMod;

    #[rstest]
    fn test_counting_intervals() {
        use std::time::Duration;

        // move the clock forward so we can go backwards below
        TestInstant::sleep(Duration::from_secs(3600));
        let now = TestInstant::now();

        let d10 = Duration::from_secs(10);
        assert_eq!(
            CrashFreeIntervalTracker::full_interval_elapsed_since(d10, &now),
            TimeMod {
                count: 0,
                mark: now
            }
        );
        assert_eq!(
            CrashFreeIntervalTracker::full_interval_elapsed_since(
                d10,
                &(now - Duration::from_secs(10))
            ),
            TimeMod {
                count: 1,
                mark: now
            }
        );
        assert_eq!(
            CrashFreeIntervalTracker::full_interval_elapsed_since(
                d10,
                &(now - Duration::from_secs(25))
            ),
            TimeMod {
                count: 2,
                mark: now - Duration::from_secs(5)
            }
        );
    }

    #[rstest]
    fn test_counting_hours() {
        let mut crashfree_tracker = CrashFreeIntervalTracker::<TestInstant>::new_hourly();

        TestInstant::sleep(Duration::from_secs(7200));

        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            2,
            2,
        );
    }

    #[rstest]
    fn test_counting_minutes() {
        let mut crashfree_tracker = CrashFreeIntervalTracker::<TestInstant>::new(
            Duration::from_secs(60),
            METRIC_OPERATIONAL_HOURS.parse().unwrap(),
            METRIC_OPERATIONAL_CRASHFREE_HOURS.parse().unwrap(),
            METRIC_OPERATIONAL_CRASHES.parse().unwrap(),
        );

        TestInstant::sleep(Duration::from_secs(3600));
        crashfree_tracker.capture_crash();
        TestInstant::sleep(Duration::from_secs(3600));

        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            120,
            60,
        );
    }

    #[rstest]
    fn test_30min_heartbeat() {
        let mut crashfree_tracker = CrashFreeIntervalTracker::<TestInstant>::new_hourly();

        TestInstant::sleep(Duration::from_secs(1800));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            0,
            0,
        );

        TestInstant::sleep(Duration::from_secs(1800));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            1,
            1,
        );
    }

    #[rstest]
    fn test_30min_heartbeat_with_crash() {
        let mut crashfree_tracker = CrashFreeIntervalTracker::<TestInstant>::new_hourly();

        TestInstant::sleep(Duration::from_secs(1800));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            0,
            0,
        );

        // Crash at t0 + 30min
        crashfree_tracker.capture_crash();

        // After 30' we should be ready to mark an operational hour
        TestInstant::sleep(Duration::from_secs(1800));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            1,
            0,
        );

        // After another 30' we should be ready to mark another crashfree hour
        TestInstant::sleep(Duration::from_secs(1800));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            0,
            1,
        );

        // After another 30' we should be ready to mark another operational hour
        TestInstant::sleep(Duration::from_secs(1800));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            1,
            0,
        );
    }

    #[rstest]
    fn test_180min_heartbeat_with_one_crash() {
        let mut crashfree_tracker = CrashFreeIntervalTracker::<TestInstant>::new_hourly();

        // Basic test
        TestInstant::sleep(Duration::from_secs(3600 * 3));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            3,
            3,
        );

        // Crash at interval + 170'
        TestInstant::sleep(Duration::from_secs(170 * 60));
        crashfree_tracker.capture_crash();

        // Another 10' to the heartbeat mark
        // We will count 0 operational hour here. That is a consequence of the heartbeat being larger than the hour
        // To avoid this bug, we need to make sure we call the `update` at least once per hour!
        TestInstant::sleep(Duration::from_secs(10 * 60));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            3,
            0,
        );

        // However, doing the crash at interval +10' then waiting for 170' will record 2 crashfree hours
        TestInstant::sleep(Duration::from_secs(10 * 60));
        crashfree_tracker.capture_crash();
        TestInstant::sleep(Duration::from_secs(170 * 60));
        assert_operational_metrics(
            crashfree_tracker.wait_and_update(Duration::from_secs(0)),
            3,
            2,
        );
    }

    fn assert_operational_metrics(
        metrics: Vec<KeyedMetricReading>,
        expected_op_hours: u32,
        expected_crashfree_hours: u32,
    ) {
        assert_eq!(metrics.len(), 3);
        let op_hours = metrics
            .iter()
            .find(|m| m.name.as_str() == METRIC_OPERATIONAL_HOURS)
            .unwrap();
        let crash_free_hours = metrics
            .iter()
            .find(|m| m.name.as_str() == METRIC_OPERATIONAL_CRASHFREE_HOURS)
            .unwrap();

        let op_hours_value = match op_hours.value {
            MetricReading::Counter { value, .. } => value,
            _ => panic!("Unexpected metric type"),
        };

        let crashfree_hours_value = match crash_free_hours.value {
            MetricReading::Counter { value, .. } => value,
            _ => panic!("Unexpected metric type"),
        };

        assert_eq!(
            (op_hours_value as u32, crashfree_hours_value as u32),
            (expected_op_hours, expected_crashfree_hours)
        );
    }
}
