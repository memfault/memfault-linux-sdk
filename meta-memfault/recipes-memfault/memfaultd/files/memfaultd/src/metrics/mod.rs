//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use chrono::{DateTime, Utc};

mod battery;
pub use battery::BatteryMonitor;
pub use battery::BatteryMonitorReading;
pub use battery::BatteryReadingHandler;

mod connectivity;
pub use connectivity::ConnectivityMonitor;
pub use connectivity::ReportSyncEventHandler;

mod metric_string_key;
pub use metric_string_key::MetricStringKey;

mod metric_report;
pub use metric_report::MetricReport;
pub use metric_report::MetricReportType;

mod metric_report_manager;
pub use metric_report_manager::MetricReportManager;
pub use metric_report_manager::TakeMetrics;

mod messages;
pub use messages::*;

mod metric_reading;
pub use metric_reading::KeyedMetricReading;
pub use metric_reading::MetricReading;

mod timeseries;

mod metric_value;
pub use metric_value::MetricValue;

pub type MetricTimestamp = DateTime<Utc>;

mod periodic_metric_report;
pub use periodic_metric_report::PeriodicMetricReportDumper;

mod crashfree_interval;
pub use crashfree_interval::CrashFreeIntervalTracker;

mod session_name;
pub use session_name::SessionName;

mod session_event_handler;
pub use session_event_handler::SessionEventHandler;

pub mod core_metrics;

pub mod statsd_server;
pub use statsd_server::StatsDServer;

mod system_metrics;
pub use system_metrics::SystemMetricsCollector;
pub use system_metrics::BUILTIN_SYSTEM_METRIC_NAMESPACES;
