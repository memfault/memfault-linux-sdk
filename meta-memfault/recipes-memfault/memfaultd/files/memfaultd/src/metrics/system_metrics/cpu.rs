//
// Copyright (c) Memfault, Inc.
// See License.txt for details
//! Collect CPU metric readings from /proc/stat
//!
//! This module parses CPU statistics from /proc/stat and
//! constructs KeyedMetricReadings based on those statistics.
//! Because the /proc/stat values are accumulations since boot,
//! a "previous reading" (stored in CpuMetricCollector) is
//! required to calculate the time each CPU core
//! has spent in each state since the last reading.
//!
//! Example /proc/stat contents:
//! cpu  326218 0 178980 36612114 6054 0 11961 0 0 0
//! cpu0 77186 0 73689 9126238 1353 0 6352 0 0 0
//! cpu1 83902 0 35260 9161039 1524 0 1865 0 0 0
//! cpu2 83599 0 35323 9161010 1676 0 1875 0 0 0
//! cpu3 81530 0 34707 9163825 1500 0 1867 0 0 0
//! intr 95400676 0 9795 1436573 0 0 0 0 0 0 0 0 93204555 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 77883 0 530 0 0 1523 0 0 468762 0 0 97412 103573 0 70 0 0 0 0 0
//! ctxt 9591503
//! btime 1714309294
//! processes 9416
//! procs_running 1
//! procs_blocked 0
//! softirq 47765068 15 3173702 0 541726 82192 0 1979 41497887 0 2467567
//!
//! Only the lines that start with "cpu" are currently
//! processed into metric readings by this module - the rest are discarded.
//!
//! See additional Linux kernel documentation on /proc/stat here:
//! https://docs.kernel.org/filesystems/proc.html#miscellaneous-kernel-statistics-in-proc-stat
use std::collections::HashMap;
use std::fs::File;
use std::io::{BufRead, BufReader};
use std::iter::zip;
use std::ops::Add;
use std::path::Path;
use std::str::FromStr;

use chrono::Utc;
use nom::{
    bytes::complete::tag,
    character::complete::{digit0, space1},
    multi::count,
    number::complete::double,
    sequence::{pair, preceded},
    IResult,
};

use crate::metrics::{KeyedMetricReading, MetricReading, MetricStringKey};
use eyre::{eyre, ErrReport, Result};

const PROC_STAT_PATH: &str = "/proc/stat";
pub const CPU_METRIC_NAMESPACE: &str = "cpu";

pub struct CpuMetricCollector {
    last_reading_by_cpu: HashMap<String, Vec<f64>>,
}

impl CpuMetricCollector {
    pub fn new() -> Self {
        Self {
            last_reading_by_cpu: HashMap::new(),
        }
    }

    pub fn get_cpu_metrics(&mut self) -> Result<Vec<KeyedMetricReading>> {
        // Track if any lines in /proc/stat are parse-able
        // so we can alert user if none are
        let mut no_parseable_lines = true;

        let path = Path::new(PROC_STAT_PATH);

        let file = File::open(path)?;
        let reader = BufReader::new(file);

        let mut cpu_metric_readings = vec![];

        for line in reader.lines() {
            // Discard errors - the assumption here is that we are only parsing
            // lines that follow the specified format and expect other lines in the file to error
            if let Ok((cpu_id, cpu_stats)) = Self::parse_proc_stat_line_cpu(line?.trim()) {
                no_parseable_lines = false;
                if let Ok(Some(mut readings)) = self.delta_since_last_reading(cpu_id, cpu_stats) {
                    cpu_metric_readings.append(&mut readings);
                }
            }
        }

        // Check if we were able to parse at least one CPU metric reading
        if !no_parseable_lines {
            Ok(cpu_metric_readings)
        } else {
            Err(eyre!(
                    "No CPU metrics were collected from {} - is it a properly formatted /proc/stat file?",
                    PROC_STAT_PATH
            ))
        }
    }

    /// Parse a cpu ID from a line of /proc/stat
    ///
    /// A cpu ID is the digit following the 3 character string "cpu"
    /// No ID being present implies these stats are for the total
    /// of all cores on the CPU
    fn parse_cpu_id(input: &str) -> IResult<&str, &str> {
        preceded(tag("cpu"), digit0)(input)
    }

    /// Parse the CPU stats from the suffix of a /proc/stat line following the cpu ID
    ///
    /// 7 or more space delimited integers are expected. Values after the 7th are discarded.
    fn parse_cpu_stats(input: &str) -> IResult<&str, Vec<f64>> {
        count(preceded(space1, double), 7)(input)
    }

    /// Parse the output of a line of /proc/stat, returning
    /// a pair of the cpu ID that the parsed line corresponds
    /// to and the first 7 floats listed for it
    ///
    /// The 7 floats represent how much time since boot the cpu has
    /// spent in the "user", "nice", "system", "idle", "iowait", "irq",
    /// "softirq", in that order    
    /// Example of a valid parse-able line:
    /// cpu2 36675 176 11216 1552961 689 0 54
    fn parse_proc_stat_line_cpu(line: &str) -> Result<(String, Vec<f64>)> {
        let (_remaining, (cpu_id, cpu_stats)) =
            pair(Self::parse_cpu_id, Self::parse_cpu_stats)(line)
                .map_err(|_e| eyre!("Failed to parse CPU stats line: {}", line))?;
        Ok(("cpu".to_string().add(cpu_id), cpu_stats))
    }

    /// Calculate the time spent in each state for the
    /// provided CPU core since the last reading collected
    /// by the CpuMetricCollector
    ///
    /// Returns an Ok(None) if there is no prior reading
    /// to calculate a delta from.
    fn delta_since_last_reading(
        &mut self,
        cpu_id: String,
        cpu_stats: Vec<f64>,
    ) -> Result<Option<Vec<KeyedMetricReading>>> {
        // Check to make sure there was a previous reading to calculate a delta with
        if let Some(last_stats) = self
            .last_reading_by_cpu
            .insert(cpu_id.clone(), cpu_stats.clone())
        {
            let delta = cpu_stats
                .iter()
                .zip(last_stats)
                .map(|(current, previous)| current - previous);

            let cpu_states_with_ticks = zip(
                ["user", "nice", "system", "idle", "iowait", "irq", "softirq"],
                delta,
            )
            .collect::<Vec<(&str, f64)>>();

            let sum: f64 = cpu_states_with_ticks.iter().map(|(_k, v)| v).sum();
            let timestamp = Utc::now();

            let readings = cpu_states_with_ticks
                .iter()
                .map(|(key, value)| -> Result<KeyedMetricReading, ErrReport> {
                    Ok(KeyedMetricReading::new(
                        MetricStringKey::from_str(&format!(
                            "{}/{}/percent/{}",
                            CPU_METRIC_NAMESPACE, cpu_id, key
                        ))
                        .map_err(|e| eyre!(e))?,
                        MetricReading::Histogram {
                            // Transform raw tick value to a percentage
                            value: 100.0 * value / sum,
                            timestamp,
                        },
                    ))
                })
                .collect::<Result<Vec<KeyedMetricReading>>>();
            match readings {
                Ok(readings) => Ok(Some(readings)),
                Err(e) => Err(e),
            }
        } else {
            Ok(None)
        }
    }
}

#[cfg(test)]
mod test {

    use insta::{assert_json_snapshot, rounded_redaction, with_settings};
    use rstest::rstest;

    use super::*;

    #[rstest]
    #[case("cpu 1000 5 0 0 2 0 0", "test_basic_line")]
    #[case("cpu0 1000 5 0 0 2 0 0 0 0 0", "test_basic_line_with_extra")]
    fn test_process_valid_proc_stat_line(#[case] proc_stat_line: &str, #[case] test_name: &str) {
        assert_json_snapshot!(test_name, 
                              CpuMetricCollector::parse_proc_stat_line_cpu(proc_stat_line).unwrap(), 
                              {"[].value.**.timestamp" => "[timestamp]", "[].value.**.value" => rounded_redaction(5)})
    }

    #[rstest]
    #[case("cpu 1000 5 0 0 2")]
    #[case("1000 5 0 0 2 0 0 0 0 0")]
    #[case("processor0 1000 5 0 0 2 0 0 0 0 0")]
    #[case("softirq 403453672 10204651 21667771 199 12328940 529390 0 3519783 161759969 147995 193294974")]
    fn test_fails_on_invalid_proc_stat_line(#[case] proc_stat_line: &str) {
        assert!(CpuMetricCollector::parse_proc_stat_line_cpu(proc_stat_line).is_err())
    }

    #[rstest]
    #[case(
        "cpu 1000 5 0 0 2 0 0",
        "cpu 1500 20 4 1 2 0 0",
        "cpu 1550 200 40 3 3 0 0",
        "basic_delta"
    )]
    fn test_cpu_metric_collector_calcs(
        #[case] proc_stat_line_a: &str,
        #[case] proc_stat_line_b: &str,
        #[case] proc_stat_line_c: &str,
        #[case] test_name: &str,
    ) {
        let mut cpu_metric_collector = CpuMetricCollector::new();

        let (cpu, stats) = CpuMetricCollector::parse_proc_stat_line_cpu(proc_stat_line_a).unwrap();
        let result_a = cpu_metric_collector.delta_since_last_reading(cpu, stats);
        matches!(result_a, Ok(None));

        let (cpu, stats) = CpuMetricCollector::parse_proc_stat_line_cpu(proc_stat_line_b).unwrap();
        let result_b = cpu_metric_collector.delta_since_last_reading(cpu, stats);

        assert!(result_b.is_ok());

        with_settings!({sort_maps => true}, {
            assert_json_snapshot!(format!("{}_{}", test_name, "a_b_metrics"),
                                  result_b.unwrap(),
                                  {"[].value.**.timestamp" => "[timestamp]", "[].value.**.value" => rounded_redaction(5)})
        });

        let (cpu, stats) = CpuMetricCollector::parse_proc_stat_line_cpu(proc_stat_line_c).unwrap();
        let result_c = cpu_metric_collector.delta_since_last_reading(cpu, stats);

        assert!(result_c.is_ok());

        with_settings!({sort_maps => true}, {
            assert_json_snapshot!(format!("{}_{}", test_name, "b_c_metrics"),
                                  result_c.unwrap(),
                                  {"[].value.**.timestamp" => "[timestamp]", "[].value.**.value" => rounded_redaction(5)})
        });
    }

    #[rstest]
    #[case(
        "cpu1 40 20 30 10 0 0 0",
        "cpu0 1500 20 4 1 2 0 0",
        "cpu1 110 30 40 12 5 3 0",
        "different_cores"
    )]
    fn test_cpu_metric_collector_different_cores(
        #[case] proc_stat_line_a: &str,
        #[case] proc_stat_line_b: &str,
        #[case] proc_stat_line_c: &str,
        #[case] test_name: &str,
    ) {
        let mut cpu_metric_collector = CpuMetricCollector::new();

        let (cpu, stats) = CpuMetricCollector::parse_proc_stat_line_cpu(proc_stat_line_a).unwrap();
        let result_a = cpu_metric_collector.delta_since_last_reading(cpu, stats);
        matches!(result_a, Ok(None));

        let (cpu, stats) = CpuMetricCollector::parse_proc_stat_line_cpu(proc_stat_line_b).unwrap();
        let result_b = cpu_metric_collector.delta_since_last_reading(cpu, stats);
        matches!(result_b, Ok(None));

        let (cpu, stats) = CpuMetricCollector::parse_proc_stat_line_cpu(proc_stat_line_c).unwrap();
        let result_c = cpu_metric_collector.delta_since_last_reading(cpu, stats);

        assert!(result_c.is_ok());

        with_settings!({sort_maps => true}, {
            assert_json_snapshot!(format!("{}_{}", test_name, "a_c_metrics"),
                                  result_c.unwrap(),
                                  {"[].value.**.timestamp" => "[timestamp]", "[].value.**.value" => rounded_redaction(5)})
        });
    }
}
