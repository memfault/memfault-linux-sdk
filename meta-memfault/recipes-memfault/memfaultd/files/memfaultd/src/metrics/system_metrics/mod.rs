//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::{
    thread::sleep,
    time::{Duration, Instant},
};

use eyre::Result;
use log::{debug, warn};

use crate::{
    config::SystemMetricConfig,
    metrics::KeyedMetricReading,
    util::system::{bytes_per_page, clock_ticks_per_second},
};

mod cpu;
use crate::metrics::system_metrics::cpu::{CpuMetricCollector, CPU_METRIC_NAMESPACE};

mod thermal;
use crate::metrics::system_metrics::thermal::{ThermalMetricsCollector, THERMAL_METRIC_NAMESPACE};

mod memory;
use crate::metrics::system_metrics::memory::{MemoryMetricsCollector, MEMORY_METRIC_NAMESPACE};

mod network_interfaces;
use crate::metrics::system_metrics::network_interfaces::{
    NetworkInterfaceMetricCollector, NetworkInterfaceMetricsConfig,
    NETWORK_INTERFACE_METRIC_NAMESPACE,
};

mod processes;
use processes::{ProcessMetricsCollector, PROCESSES_METRIC_NAMESPACE};

mod disk_space;
use disk_space::{
    DiskSpaceMetricCollector, DiskSpaceMetricsConfig, NixStatvfs, DISKSPACE_METRIC_NAMESPACE,
    DISKSPACE_METRIC_NAMESPACE_LEGACY,
};

use self::processes::ProcessMetricsConfig;
use super::MetricsMBox;

pub const BUILTIN_SYSTEM_METRIC_NAMESPACES: &[&str; 7] = &[
    CPU_METRIC_NAMESPACE,
    MEMORY_METRIC_NAMESPACE,
    THERMAL_METRIC_NAMESPACE,
    NETWORK_INTERFACE_METRIC_NAMESPACE,
    PROCESSES_METRIC_NAMESPACE,
    DISKSPACE_METRIC_NAMESPACE,
    // Include in list of namespaces so that
    // legacy collectd from the "df" plugin
    // are still filtered out
    DISKSPACE_METRIC_NAMESPACE_LEGACY,
];

pub trait SystemMetricFamilyCollector {
    fn collect_metrics(&mut self) -> Result<Vec<KeyedMetricReading>>;
    fn family_name(&self) -> &'static str;
}

pub struct SystemMetricsCollector {
    metric_family_collectors: Vec<Box<dyn SystemMetricFamilyCollector>>,
    metrics_mbox: MetricsMBox,
}

impl SystemMetricsCollector {
    pub fn new(system_metric_config: SystemMetricConfig, metrics_mbox: MetricsMBox) -> Self {
        // CPU, Memory, and Thermal metrics are captured by default
        let mut metric_family_collectors: Vec<Box<dyn SystemMetricFamilyCollector>> = vec![
            Box::new(CpuMetricCollector::new()),
            Box::new(MemoryMetricsCollector::new()),
            Box::new(ThermalMetricsCollector::new()),
        ];

        // Check if process metrics have been manually configured
        match system_metric_config.processes {
            Some(processes) if !processes.is_empty() => {
                metric_family_collectors.push(Box::new(ProcessMetricsCollector::<Instant>::new(
                    ProcessMetricsConfig::Processes(processes),
                    clock_ticks_per_second() as f64 / 1000.0,
                    bytes_per_page() as f64,
                )))
            }
            // Monitoring no processes means this collector is disabled
            Some(_empty_set) => {}
            None => {
                metric_family_collectors.push(Box::new(ProcessMetricsCollector::<Instant>::new(
                    ProcessMetricsConfig::Auto,
                    clock_ticks_per_second() as f64 / 1000.0,
                    bytes_per_page() as f64,
                )))
            }
        };

        // Check if disk space metrics have been manually configured
        match system_metric_config.disk_space {
            Some(disks) if !disks.is_empty() => {
                metric_family_collectors.push(Box::new(DiskSpaceMetricCollector::new(
                    NixStatvfs::new(),
                    DiskSpaceMetricsConfig::Disks(disks),
                )))
            }
            // Monitoring no disks means this collector is disabled
            Some(_empty_set) => {}
            None => metric_family_collectors.push(Box::new(DiskSpaceMetricCollector::new(
                NixStatvfs::new(),
                DiskSpaceMetricsConfig::Auto,
            ))),
        };

        // Check if network interface metrics have been manually configured
        match system_metric_config.network_interfaces {
            Some(interfaces) if !interfaces.is_empty() => metric_family_collectors.push(Box::new(
                NetworkInterfaceMetricCollector::<Instant>::new(
                    NetworkInterfaceMetricsConfig::Interfaces(interfaces),
                ),
            )),
            // Monitoring no interfaces means this collector is disabled
            Some(_empty_set) => {}
            None => metric_family_collectors.push(Box::new(NetworkInterfaceMetricCollector::<
                Instant,
            >::new(
                NetworkInterfaceMetricsConfig::Auto
            ))),
        };

        Self {
            metric_family_collectors,
            metrics_mbox,
        }
    }

    pub fn run(&mut self, metric_poll_duration: Duration) {
        loop {
            for collector in self.metric_family_collectors.iter_mut() {
                match collector.collect_metrics() {
                    Ok(readings) => {
                        if let Err(e) = self.metrics_mbox.send_and_forget(readings) {
                            debug!(
                                "Couldn't add metric reading for family \"{}\": {:?}",
                                collector.family_name(),
                                e
                            )
                        }
                    }
                    Err(e) => warn!(
                        "Failed to collect readings for family \"{}\": {}",
                        collector.family_name(),
                        e
                    ),
                }
            }

            sleep(metric_poll_duration);
        }
    }
}
