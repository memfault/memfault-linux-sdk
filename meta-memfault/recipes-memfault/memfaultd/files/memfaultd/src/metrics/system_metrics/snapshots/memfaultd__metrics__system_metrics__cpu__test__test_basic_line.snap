---
source: memfaultd/src/metrics/system_metrics/cpu.rs
expression: "CpuMetricCollector::parse_proc_stat_line_cpu(proc_stat_line).unwrap()"
---
[
  "cpu",
  [
    1000.0,
    5.0,
    0.0,
    0.0,
    2.0,
    0.0,
    0.0
  ]
]
